server/tests/raft_metadata_cluster.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use bytestring::ByteString;
+use enumset::EnumSet;
+use googletest::prelude::err;
+use googletest::{assert_that, pat, IntoTestResult};
+use rand::prelude::SliceRandom;
+use restate_core::metadata_store::{Precondition, WriteError};
+use restate_core::{cancellation_watcher, TaskCenter, TaskKind};
+use restate_local_cluster_runner::cluster::Cluster;
+use restate_local_cluster_runner::node::{BinarySource, HealthCheck, Node};
+use restate_metadata_server::create_client;
+use restate_metadata_server::tests::Value;
+use restate_types::config::{
+    Configuration, MetadataServerKind, MetadataStoreClient, MetadataStoreClientOptions, RaftOptions,
+};
+use restate_types::Versioned;
+use std::time::{Duration, Instant};
+use tracing::info;
+
+#[test_log::test(restate_core::test)]
+async fn raft_metadata_cluster_smoke_test() -> googletest::Result<()> {
+    let base_config = Configuration::default();
+
+    let nodes = Node::new_test_nodes(
+        base_config,
+        BinarySource::CargoTest,
+        EnumSet::empty(),
+        3,
+        true,
+    );
+    let mut cluster = Cluster::builder()
+        .cluster_name("raft_metadata_cluster_smoke_test")
+        .nodes(nodes)
+        .temp_base_dir()
+        .build()
+        .start()
+        .await?;
+
+    cluster.wait_healthy(Duration::from_secs(30)).await?;
+
+    let addresses = cluster
+        .nodes
+        .iter()
+        .map(|node| node.node_address().clone())
+        .collect();
+
+    let metadata_store_client_options = MetadataStoreClientOptions {
+        metadata_store_client: MetadataStoreClient::Embedded { addresses },
+        ..MetadataStoreClientOptions::default()
+    };
+    let client = create_client(metadata_store_client_options)
+        .await
+        .expect("to not fail");
+
+    let value = Value::new(42);
+    let value_version = value.version();
+    let key = ByteString::from_static("my-key");
+    client
+        .put(key.clone(), &value, Precondition::DoesNotExist)
+        .await?;
+
+    let stored_value = client.get::<Value>(key.clone()).await?;
+    assert_eq!(stored_value, Some(value));
+
+    let stored_value_version = client.get_version(key.clone()).await?;
+    assert_eq!(stored_value_version, Some(value_version));
+
+    let new_value = Value::new(1337);
+    let new_value_version = new_value.version();
+    let failed_precondition = client
+        .put(
+            key.clone(),
+            &new_value,
+            Precondition::MatchesVersion(value_version.next()),
+        )
```

Comment:
This is perfect for a smoke test but if you wanted to, you could also add a `Precondition::DoesNotExist` attempt. Maybe it just makes the test extra verbose though, not sure.

Reply:
Good idea. Will add it as well.